lib/auth/recordingmetadata/recordingmetadatav1/recordingmetadata.go

```diff
@@ -0,0 +1,350 @@
+	defer cancel()
+
+	evts, errors := s.streamer.StreamSessionEvents(ctx, sessionID, 0)
+
+	var startTime time.Time
+	var lastEvent apievents.AuditEvent
+	var lastActivityTime time.Time
+
+	thumbnailInterval := 1 * time.Second
+	activeUsers := make(map[string]time.Duration)
+
+	vt := vt10x.New()
+
+	metadata := &pb.SessionRecordingMetadata{
+		Events: make([]*pb.SessionRecordingEvent, 0),
+	}
+
+	addInactivityEvent := func(start, end time.Time) {
+		if end.IsZero() {
+			return
+		}
+
+		inactivityStart := durationpb.New(start.Sub(startTime))
+		inactivityEnd := durationpb.New(end.Sub(startTime))
+
+		metadata.Events = append(metadata.Events, &pb.SessionRecordingEvent{
+			StartOffset: inactivityStart,
+			EndOffset:   inactivityEnd,
+			Event: &pb.SessionRecordingEvent_Inactivity{
+				Inactivity: &pb.SessionRecordingInactivityEvent{},
+			},
+		})
+	}
+
+	sampler := newThumbnailBucketSampler(maxThumbnails, thumbnailInterval)
+
+	recordThumbnail := func(start time.Time) {
+		state := vt.DumpState()
+
+		sampler.add(&state, start)
+	}
+
+loop:
+	for {
+		select {
+		case evt, ok := <-evts:
+			if !ok {
+				break loop
+			}
+
+			lastEvent = evt
+
+			switch e := evt.(type) {
+			case *apievents.DatabaseSessionStart, *apievents.WindowsDesktopSessionStart:
+				// Unsupported session recording types
+				return nil
+
+			case *apievents.Resize:
+				size, err := session.UnmarshalTerminalParams(e.TerminalSize)
+				if err != nil {
+					return trace.Wrap(err, "failed to parse terminal size %q for session %v", e.TerminalSize, sessionID)
+				}
+
+				metadata.Events = append(metadata.Events, &pb.SessionRecordingEvent{
+					StartOffset: durationpb.New(e.Time.Sub(startTime)),
+					Event: &pb.SessionRecordingEvent_Resize{
+						Resize: &pb.SessionRecordingResizeEvent{
+							Cols: int32(size.W),
+							Rows: int32(size.H),
+						},
+					},
+				})
+
+				vt.Resize(size.W, size.H)
+
+			case *apievents.SessionEnd:
+				if !lastActivityTime.IsZero() && e.Time.Sub(lastActivityTime) > inactivityThreshold {
+					addInactivityEvent(lastActivityTime, e.Time)
+				}
+
+				recordThumbnail(e.EndTime)
+
+			case *apievents.SessionJoin:
+				activeUsers[e.User] = e.Time.Sub(startTime)
+
+			case *apievents.SessionLeave:
+				if joinTime, ok := activeUsers[e.User]; ok {
+					metadata.Events = append(metadata.Events, &pb.SessionRecordingEvent{
+						StartOffset: durationpb.New(joinTime),
+						EndOffset:   durationpb.New(e.Time.Sub(startTime)),
+						Event: &pb.SessionRecordingEvent_Join{
+							Join: &pb.SessionRecordingJoinEvent{
+								User: e.User,
+							},
+						},
+					})
+
+					delete(activeUsers, e.User)
+				}
+
+			case *apievents.SessionPrint:
+				if !lastActivityTime.IsZero() && e.Time.Sub(lastActivityTime) > inactivityThreshold {
+					addInactivityEvent(lastActivityTime, e.Time)
+				}
+
+				if _, err := vt.Write(e.Data); err != nil {
+					return trace.Errorf("failed to write data to terminal: %w", err)
```

Comment:
Nit: when wrapping errors in Go, I tend to prefer avoiding words like "failed" or "error" as when it's been wrapped a few times, you end up with a message that repeats those words.

```suggestion
					return trace.Errorf("writing data to terminal: %w", err)
```